flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/planner/Planner.java

```diff
@@ -0,0 +1,99 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.streaming.api.transformations.StreamTransformation;
+import org.apache.flink.table.operations.ModifyOperation;
+import org.apache.flink.table.operations.QueryOperation;
+import org.apache.flink.table.operations.TableOperation;
+
+import java.util.List;
+
+/**
+ * This interface serves two purposes:
+ * <ul>
+ * <li>SQL parser - transforms a SQL string into a Table API specific tree of
+ * {@link TableOperation}s</li>
+ * <li>relational planner - provides a way to plan, optimize and transform tree of
+ * {@link ModifyOperation} into a runnable form ({@link StreamTransformation})</li>
+ * </ul>.
+ *
+ * <p>The Planner is execution agnostic. It is up to the
+ * {@link org.apache.flink.table.api.TableEnvironment} to ensure that if any of the
+ * {@link QueryOperation} pull any runtime configuration, all those configurations are
+ * equivalent. Example: If some of the {@link QueryOperation}s scan DataStreams, all
+ * those DataStreams must come from the same StreamExecutionEnvironment, because the result
+ * of {@link Planner#translate(List)} will strip any execution configuration from
+ * the DataStream information.
+ *
+ * <p>All Tables referenced in either {@link Planner#parse(String)} or
+ * {@link Planner#translate(List)} should be previously registered in a
+ * {@link org.apache.flink.table.catalog.CatalogManager}, which will be provided during
+ * instantiation of the {@link Planner}.
+ */
+@Internal
+public interface Planner {
+
+	/**
+	 * Entry point for parsing sql queries expressed as a String.
+	 *
+	 * <p><b>Note:</b>If the created {@link TableOperation} is a {@link QueryOperation}
+	 * it must be in a form that will be understood by the
+	 * {@link Planner#translate(List)} method.
+	 *
+	 * <p>The produced Operation tree should already be validated.
+	 *
+	 * @param stmt the sql statement to evaluate
+	 * @return parsed query as a tree of relational {@link TableOperation}s
+	 */
+	TableOperation parse(String stmt);
+
+	/**
+	 * Converts a relational tree of {@link QueryOperation}s into a set of runnable
+	 * {@link StreamTransformation}s.
+	 *
+	 * <p>This method accepts a list of {@link QueryOperation}s to allow reusing common
+	 * subtrees of multiple relational queries.
+	 *
+	 * @param tableOperations list of relational operations to plan, optimize and convert in a
+	 * single run.
+	 * @return list of corresponding {@link StreamTransformation}s.
+	 */
+	List<StreamTransformation<?>> translate(List<ModifyOperation> tableOperations);
```

Comment:
the argument and the comments are inconsistent: the argument is a list of `ModifyOperation`, while the comments describe a list of `QueryOperation`. 

btw, if this method accepts a list of `ModifyOperation`, do we need implicitly add sink or output conversion for plan test queries? like queries in `CalcTest`

Reply:
Sorry for the confusion. I updated the javadoc. Yes it should be always `ModifyOperation` as this is the way we can pass such configuration as update mode or expected output type.

I think blink does a similar thing internally with `DataStreamTableSink`.

Which `CalcTest` do you have in mind? If you are using `TableEnvironment` then it will be added there. If you have test utilities than I guess you should update it.

Reply:
I get your point.  

The `CalcTest` i mentioned above are `org.apache.flink.table.api.batch.sql.CalcTest` in flink planner and `org.apache.flink.table.plan.batch.sql.CalcTest` in blink planner.